File: training/llmviz_train/kan.py
"""Minimal KAN (Kolmogorov–Arnold) reference with activation recording.

A GPT whose feed-forward is a KAN instead of an MLP. The distinctive KAN
idea is a learnable univariate function on every edge; classic KANs use
B-splines, but those are awkward to port exactly, so this uses the FastKAN
radial-basis form (Li 2024): each edge function is a weighted sum of fixed
Gaussian bumps on a grid, which is a plain `exp` and is trivial to mirror
in TS at 1e-4.

KAN feed-forward over a pre-LayerNorm'd input x̂ (d_model -> d_model):
    rbf  = exp(-((x̂[...,None] - grid) / denom)^2)     # (T, d_model, G)
    out  = spline_linear(rbf.flatten) + base_linear(silu(x̂))

Attention is exactly the GPT CausalSelfAttention (imported, not re-derived).
"""

import torch
import torch.nn.functional as F
from torch import nn

from llmviz_train.gpt import CausalSelfAttention
from llmviz_train.recording import snapshot


class KANLayer(nn.Module):
    """FastKAN feed-forward (no residual; the model adds it). d_in == d_out."""

    def __init__(self, d_model: int, num_grids: int, grid_min: float, grid_max: float):
        super().__init__()
        self.d_model = d_model
        self.num_grids = num_grids
        grid = torch.linspace(grid_min, grid_max, num_grids)
        self.register_buffer("grid", grid, persistent=False)  # fixed, not exported
        self.denom = (grid_max - grid_min) / (num_grids - 1)
        self.spline_linear = nn.Linear(d_model * num_grids, d_model, bias=False)
        self.base_linear = nn.Linear(d_model, d_model, bias=True)

    def forward(
        self, x: torch.Tensor, record: dict[str, torch.Tensor] | None = None, prefix: str = ""
    ) -> torch.Tensor:
        """x: (B, T, d_model) (already LayerNorm'd) -> (B, T, d_model)."""

        def rec(name: str, tensor: torch.Tensor) -> None:
            if record is not None:
                record[prefix + name] = snapshot(tensor)

        # Gaussian basis on each scalar input: (B, T, d_model, G)
        rbf = torch.exp(-(((x[..., None] - self.grid) / self.denom) ** 2))
        flat = rbf.reshape(*x.shape[:-1], self.d_model * self.num_grids)
        rec("rbf", flat[0])  # (T, d_model*G) — the learnable edge functions, evaluated
        spline = self.spline_linear(flat)
        rec("spline", spline[0])
        base = self.base_linear(F.silu(x))
        rec("base", base[0])
        out = spline + base
        rec("out", out[0])
        return out


class KAN(nn.Module):
    """Minimal KAN-GPT language model over a small token vocabulary."""

    def __init__(self, cfg: dict, vocab_size: int, max_seq_len: int = 32):
        super().__init__()
        n_layer = cfg["n_layer"]
        d_model = cfg["d_model"]
        self.max_seq_len = max_seq_len

        self.tok_embedding = nn.Embedding(vocab_size, d_model)
        self.pos_embedding = nn.Embedding(max_seq_len, d_model)
        self.ln1s = nn.ModuleList(nn.LayerNorm(d_model) for _ in range(n_layer))
        self.attns = nn.ModuleList(CausalSelfAttention(d_model, cfg["n_head"]) for _ in range(n_layer))
        self.ln2s = nn.ModuleList(nn.LayerNorm(d_model) for _ in range(n_layer))
        self.kans = nn.ModuleList(
            KANLayer(d_model, cfg["num_grids"], cfg["grid_min"], cfg["grid_max"])
            for _ in range(n_layer)
        )
        self.ln_f = nn.LayerNorm(d_model)
        self.lm_head = nn.Linear(d_model, vocab_size, bias=False)  # untied

    def forward(
        self, tokens: torch.Tensor, record: bool = False
    ) -> torch.Tensor | tuple[torch.Tensor, dict[str, torch.Tensor]]:
        """tokens: (B, T) long -> logits (B, T, vocab)."""
        T = tokens.shape[1]
        assert T <= self.max_seq_len, f"T={T} exceeds max_seq_len={self.max_seq_len}"
        acts: dict[str, torch.Tensor] | None = {} if record else None

        def rec(name: str, tensor: torch.Tensor) -> None:
            if acts is not None:
                acts[name] = snapshot(tensor)

        positions = torch.arange(T, device=tokens.device)
        x = self.tok_embedding(tokens) + self.pos_embedding(positions)
        rec("embed.out", x[0])

        for i in range(len(self.attns)):
            normed = self.ln1s[i](x)
            rec(f"layer{i}.ln1.out", normed[0])
            x = x + self.attns[i](normed, record=acts, prefix=f"layer{i}.attn.")
            normed = self.ln2s[i](x)
            rec(f"layer{i}.ln2.out", normed[0])
            x = x + self.kans[i](normed, record=acts, prefix=f"layer{i}.kan.")

        x = self.ln_f(x)
        rec("final_norm.out", x[0])

        logits = self.lm_head(x)
        rec("head.logits", logits[0])

        if acts is not None:
            return logits, acts
        return logits

File: src/compute/kan.ts
/**
 * TS port of the KAN forward pass in training/llmviz_train/kan.py,
 * operation-for-operation, recording the same activation names. Gated by
 * kan.golden.test.ts against PyTorch goldens at 1e-4.
 *
 * A GPT whose feed-forward is a FastKAN: each edge function is a weighted
 * sum of fixed Gaussian bumps on a grid. Attention is reused verbatim from
 * gpt.ts. The KAN feed-forward over a LayerNorm'd input x̂:
 *   rbf  = exp(-((x̂ − grid)/denom)²)            (T, d_model, G), flattened
 *   out  = spline_linear(rbf) + base_linear(silu(x̂))
 */

import { T } from "./tensor";
import { add, embedding, layernorm, linear, silu } from "./ops";
import { getTensor, type Manifest } from "./loader";
import type { Recorder } from "./recorder";
import { dimsFromManifest } from "./util";
import { causalSelfAttention } from "./gpt";

export interface KanDims {
  n_layer: number;
  d_model: number;
  n_head: number;
  num_grids: number;
  grid_min: number;
  grid_max: number;
  vocab_size: number;
  head_dim: number;
  max_seq_len: number;
}

const KAN_DIM_KEYS = [
  "n_layer",
  "d_model",
  "n_head",
  "num_grids",
  "grid_min",
  "grid_max",
  "vocab_size",
  "head_dim",
  "max_seq_len",
] as const;

export function kanDimsFrom(manifest: Manifest): KanDims {
  const dims = dimsFromManifest(manifest, "kan", KAN_DIM_KEYS);
  if (dims.n_head * dims.head_dim !== dims.d_model) {
    throw new Error("kan dims inconsistent: n_head·head_dim != d_model");
  }
  return dims;
}

/**
 * Flatten the Gaussian-basis evaluation of x̂ to (T, d_model·G), in the
 * (channel, grid) row-major order the spline_linear weight expects.
 */
function rbfFeatures(x: T, dims: KanDims): T {
  const [steps, d] = x.shape;
  const g = dims.num_grids;
  const denom = (dims.grid_max - dims.grid_min) / (g - 1);
  const grid = new Float32Array(g);
  for (let j = 0; j < g; j++) grid[j] = dims.grid_min + j * denom;
  const out = new Float32Array(steps * d * g);
  for (let t = 0; t < steps; t++) {
    for (let i = 0; i < d; i++) {
      const xv = x.data[t * d + i];
      const base = (t * d + i) * g;
      for (let j = 0; j < g; j++) {
        const z = (xv - grid[j]) / denom;
        out[base + j] = Math.exp(-(z * z));
      }
    }
  }
  return T.from(out, [steps, d * g]);
}

/** One KAN feed-forward (no residual). x already LayerNorm'd: (T, d_model). */
function kanBlock(
  weights: Map<string, T>,
  dims: KanDims,
  layer: number,
  x: T,
  rec: Recorder | undefined,
): T {
  const rbf = rbfFeatures(x, dims);
  rec?.record(`layer${layer}.kan.rbf`, rbf);
  const spline = linear(rbf, getTensor(weights, `kans.${layer}.spline_linear.weight`));
  rec?.record(`layer${layer}.kan.spline`, spline);
  const base = linear(
    silu(x),
    getTensor(weights, `kans.${layer}.base_linear.weight`),
    getTensor(weights, `kans.${layer}.base_linear.bias`),
  );
  rec?.record(`layer${layer}.kan.base`, base);
  const out = add(spline, base);
  rec?.record(`layer${layer}.kan.out`, out);
  return out;
}

/** Full forward: tokens → logits (T, vocab), recording the kan.py names. */
export function kanForward(
  weights: Map<string, T>,
  dims: KanDims,
  tokens: number[],
  rec?: Recorder,
): T {
  if (tokens.length > dims.max_seq_len) {
    throw new Error(`T=${tokens.length} exceeds max_seq_len=${dims.max_seq_len}`);
  }
  const positions = tokens.map((_, i) => i);
  let x = add(
    embedding(tokens, getTensor(weights, "tok_embedding.weight")),
    embedding(positions, getTensor(weights, "pos_embedding.weight")),
  );
  rec?.record("embed.out", x);

  for (let i = 0; i < dims.n_layer; i++) {
    let normed = layernorm(
      x,
      getTensor(weights, `ln1s.${i}.weight`),
      getTensor(weights, `ln1s.${i}.bias`),
    );
    rec?.record(`layer${i}.ln1.out`, normed);
    x = add(x, causalSelfAttention(weights, dims, i, normed, rec));

    normed = layernorm(
      x,
      getTensor(weights, `ln2s.${i}.weight`),
      getTensor(weights, `ln2s.${i}.bias`),
    );
    rec?.record(`layer${i}.ln2.out`, normed);
    x = add(x, kanBlock(weights, dims, i, normed, rec));
  }

  x = layernorm(x, getTensor(weights, "ln_f.weight"), getTensor(weights, "ln_f.bias"));
  rec?.record("final_norm.out", x);

  const logits = linear(x, getTensor(weights, "lm_head.weight"));
  rec?.record("head.logits", logits);
  return logits;
}
